Fix data races in the MySQL Monitor and HostGroups

Gonlo2 opened the pull request:
This change fix some data races in the MySQL Monitor and HostGroups code along with some other minor bug fixes.

### A clear description of the issue

The MySQL Monitor and HostGroups code has race conditions that can leave ProxySQL in an inconsistent state, for example removing the main node from the candidate list with no possibility of recovery until another failure is detected.

### ProxySQL version
It was detected in the ProxySQL v2.0.10 build.

### OS Version
16.04.1-Ubuntu

### The steps to reproduce the issue
Usually it is detected a few minutes after several node failures, I have left a commit with two scripts and a docker-compose to reproduce the problem in the folder "reproduce", the first one generates load in a Galera cluster through sysbench and the second one restarts one of the nodes until the problem is detected (to do this the script needs that the build is a debug one to have access to the table "myhgm.mysql_servers"). The steps to execute this are:

```
$ cd reproduce
$ docker-compose up # In some terminal
$ ./generate-load.sh # In other terminal
$ ./node-error.sh # In another terminal
```

### **Full** ProxySQL error log
I don't have a complete log of this but I guess it's more flexible to get the log from the script to reproduce the problem, if this is not enough let me know I'll add a log.

Gonlo2 committed:
Refactor of the hostgroup and monitor locks to avoid data races and avoid use non isolated proxysql admin mysql_servers

Gonlo2 committed:
Check only online nodes

Gonlo2 committed:
Mark the mysql_servers table as dirty

Gonlo2 committed:
Make atomic the ProxySQL admin load to runtime method

The load to runtime method don't check if the existing datastructures
are in use before delete them, also without a lock the hostgroup could
see a incomplete picture of the changes.

Gonlo2 committed:
Avoid keep the sqlite mysql_servers table out of sync from the internal datastructure

In some cases the mysql_servers datastructure is updated but the mysql_servers
sqlite table has the previous values, so changes are not taken into account.

Gonlo2 committed:
Fix skip one server if the remove_index_fast is called

Gonlo2 committed:
Rename unsafe functions to locked

Gonlo2 committed:
Rename cleanup function

Gonlo2 committed:
Reproduce stats

pondix commented:
Automated message: PR pending admin approval for build testing

JavierJF commented:
Hi, thanks for PR, several notes about it:

- The pull request doesn't compile:

```
MySQL_HostGroups_Manager.cpp:2857:8: error: ‘unsafe_set_mysql_servers_table_dirty’ was not declared in this scope; did you mean ‘is_mysql_servers_table_dirty’?
 2857 |        unsafe_set_mysql_servers_table_dirty();
      |        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |        is_mysql_servers_table_dirty
```
- The configuration used for ProxySQL to reproduce the issue is not supplied. I have tried to deduce it, but I haven't been able to reproduce a similar error of the one you are experiencing. If you could supply the configuration file for ProxySQL, that would be of great help.

- The issue was detected in v2.0.10, but this pull request is against v2.0.14, have you been able to reproduce the issue in v2.0.14?

- The complete error log from ProxySQL and the state of servers when you detect the inconsistency you are seeing will also help to trace the problem.

Thanks.

Gonlo2 committed:
Fix bad rebase

Gonlo2 committed:
Force git to add the proxysql configuration

Gonlo2 commented:
>     * The pull request doesn't compile:

It seems that I have made a bad rebase of the changes, I added a commit with the changes that had been left out.

>     * The configuration used for ProxySQL to reproduce the issue is not supplied. I have tried to deduce it, but I haven't been able to reproduce a similar error of the one you are experiencing. If you could supply the configuration file for ProxySQL, that would be of great help.

About this, it seems that the configuration has been left out by gitignore, so I added it in another commit.

>     * The issue was detected in v2.0.10, but this pull request is against v2.0.14, have you been able to reproduce the issue in v2.0.14?

The version we use in production and where we detected the bug is 2.0.10, so my changes are from that version. Anyway I reproduced them some time ago in 2.0.13 and seeing the code of 2.0.14 I would say that it has the same race conditions bugs.

>     * The complete error log from ProxySQL and the state of servers when you detect the inconsistency you are seeing will also help to trace the problem.

I no longer have access to the logs since I detected the bug about a month ago (we had some time validating the new version in the pre-production environment). Is it worth getting the logs of the script to reproduce the changes?

Gonlo2 commented:
Any update about this @JavierJF ?

JavierJF commented:
Hi @Gonlo2, thanks for you PR again,
I have been trying to reproduce the issue you describe with your provided scripts, I have tried with several versions of ProxySQL (including v2.0.16 and v2.0.14), for being sure I left the scripts running 45 min with each version, I haven't had any luck so far.

Secondly, this PR has too many changes in it, typically we stick to one PR, on issue, if the goal of this PR is to fix a data-race, will be much more simpler to analyze it, if only packs the absolute minimum to fix it. Also, because this is for sure a timing issue, and looks like maybe wont be easy to reproduce in our side, If you are sure that the data-race you are experiencing is solved with this PR, could you please describe more deeply which one do you think are the causes of it?, i.e:

1. Conditions that triggers it.
2. Section of the code involved in it.

This way, if we understand the exact issue that you are describing, we can also come with a way of reproducing it. Thanks you!

Gonlo2 commented:
Hi, about the PR we were having several problems using ProxySQL with Galera because it didn't correctly use the priority of the nodes after a fail-over, so after investigating it I narrowed down the problem to race conditions and after reviewing the code I discovered several problems in the code corresponding to the monitoring.

Regarding the code involved I would say that is all that appears in the review and I don't think it's easy to divide even more the review with the time I have available, basically the problems that I found are:

* The hostgroups manager code is not atomic since the state may change while a check is running, an example could be the call to the commit function, which acquires a mutex, and the subsequent acquisition of the same again without taking into account that the state may have changed in the meantime. To solve this, I added a commit_locked function and the mutex is acquired during all the check code.
* The ProxySQL admin code update and delete several sqlite results without obtaining a mutex so other threads may continue using a nonexistent structure in the meantime. To solve this I change it to obtain the mutex during the process and also I used the commit_locked function.
* In some places the administration DB is consulted in the monitoring code instead the internal one, which is incorrect and by not using mutexes it has race conditions. To solve it I changed it with the internal one.
* The code used to decide if it is necessary to update the servers does not take into account if the server is offline since it has to continue maintaining the internal DB row until it has no more connections because the only place where the pointer to it is stored is the internal DB. To solve this, I added a special list with all the servers that are offline and a variable to indicate if the mysql_servers table is dirty and it is necessary to reconstruct it.

I understand it's a big review but I think it's necessary to avoid headaches in general and especially for people who use Galera. If you have any questions, please do not hesitate to contact me.

As a note, since we started using this patch in production a couple of months ago we haven't had any more problems with fail-over in production although, unfortunately, I don't have the resources to review in depth the changes that are not Galera related.

mirostauder commented:
Can one of the admins verify this patch?

Gonlo2 commented:
Any update about this?
